diff -urN a/chrome/browser/download/download_browsertest.cc b/chrome/browser/download/download_browsertest.cc
--- a/chrome/browser/download/download_browsertest.cc	2019-10-27 01:17:57.325563389 +0000
+++ b/chrome/browser/download/download_browsertest.cc	2019-10-27 01:19:14.203034538 +0000
@@ -1256,7 +1256,7 @@
   CheckDownload(browser(), file, file);
 }
 
-#if defined(OS_WIN)
+#if defined(OS_WIN) || (defined(OS_LINUX) && !defined(OS_CHROMEOS))
 // Download a file and confirm that the file is correctly quarantined.
 //
 // TODO(asanka): We should enable the test on Mac as well, but currently
@@ -1280,7 +1280,9 @@
   EXPECT_TRUE(quarantine::IsFileQuarantined(downloaded_file, url, GURL()));
   CheckDownload(browser(), file, file);
 }
+#endif
 
+#if defined(OS_WIN)
 // A couple of Windows specific tests to make sure we respect OS specific
 // restrictions on Mark-Of-The-Web can be applied. While Chrome doesn't directly
 // apply these policies, Chrome still needs to make sure the correct APIs are
diff -urN a/chrome/test/ppapi/ppapi_filechooser_browsertest.cc b/chrome/test/ppapi/ppapi_filechooser_browsertest.cc
--- a/chrome/test/ppapi/ppapi_filechooser_browsertest.cc	2019-10-27 01:17:57.325563389 +0000
+++ b/chrome/test/ppapi/ppapi_filechooser_browsertest.cc	2019-10-27 01:20:14.047316609 +0000
@@ -236,12 +236,17 @@
   RunTestViaHTTP("FileChooser_SaveAsCancel");
 }
 
-#if defined(OS_WIN) || defined(OS_MACOSX)
-// On Windows and macOS, tests that a file downloaded via PPAPI FileChooser API
-// has the mark-of-the-web. The PPAPI FileChooser implementation invokes
-// QuarantineFile in order to mark the file as being downloaded from the web as
-// soon as the file is created. This MotW prevents the file being opened without
-// due security warnings if the file is executable.
+#if defined(OS_WIN) || defined(OS_MACOSX) || defined(OS_LINUX) && !defined(OS_CHROMEOS)
+// On Windows, tests that a file downloaded via PPAPI FileChooser API has the
+// mark-of-the-web. The PPAPI FileChooser implementation invokes QuarantineFile
+// in order to mark the file as being downloaded from the web as soon as the
+// file is created. This MOTW prevents the file being opened without due
+// security warnings if the file is executable.
+//
+// On Linux Desktop, the setxattr call is made to set 'user.xdg.origin.url' and
+// the non-standard 'user.xdg.referrer.url' extended attributes to accomplish
+// the same thing. See
+// https://www.freedesktop.org/wiki/CommonExtendedAttributes/.
 IN_PROC_BROWSER_TEST_F(PPAPIFileChooserTest, FileChooser_Quarantine) {
   base::ScopedAllowBlockingForTesting allow_blocking;
   base::ScopedTempDir temp_dir;
@@ -261,7 +266,7 @@
   ASSERT_TRUE(base::PathExists(actual_filename));
   EXPECT_TRUE(quarantine::IsFileQuarantined(actual_filename, GURL(), GURL()));
 }
-#endif  // defined(OS_WIN) || defined(OS_MACOSX)
+#endif  // defined(OS_WIN) || defined(OS_MACOSX) || defined(OS_LINUX) && !defined(OS_CHROMEOS)
 
 #if BUILDFLAG(FULL_SAFE_BROWSING)
 // These tests only make sense when SafeBrowsing is enabled. They verify
diff -urN a/components/services/quarantine/BUILD.gn b/components/services/quarantine/BUILD.gn
--- a/components/services/quarantine/BUILD.gn	2019-10-27 01:17:57.325563389 +0000
+++ b/components/services/quarantine/BUILD.gn	2019-10-27 01:21:20.125282291 +0000
@@ -14,6 +14,7 @@
     "quarantine.h",
     "quarantine_impl.cc",
     "quarantine_impl.h",
+    "quarantine_linux.cc",
     "quarantine_mac.mm",
     "quarantine_win.cc",
   ]
@@ -45,6 +46,8 @@
   sources = [
     "common.cc",
     "common.h",
+    "common_linux.cc",
+    "common_linux.h",
     "common_mac.h",
     "common_mac.mm",
     "common_win.cc",
@@ -70,6 +73,7 @@
   sources = [
     "test_support.cc",
     "test_support.h",
+    "test_support_linux.cc",
     "test_support_mac.mm",
     "test_support_win.cc",
   ]
@@ -94,6 +98,7 @@
 
   sources = [
     "common_unittests.cc",
+    "quarantine_linux_unittest.cc",
     "quarantine_mac_unittest.mm",
     "quarantine_service_unittest.cc",
     "quarantine_win_unittest.cc",
diff -urN a/components/services/quarantine/quarantine.cc b/components/services/quarantine/quarantine.cc
--- a/components/services/quarantine/quarantine.cc	2019-10-27 01:17:57.328896745 +0000
+++ b/components/services/quarantine/quarantine.cc	2019-10-27 01:21:38.329011817 +0000
@@ -6,7 +6,7 @@
 
 #include "build/build_config.h"
 
-#if !defined(OS_WIN) && !defined(OS_MACOSX)
+#if !defined(OS_WIN) && !defined(OS_MACOSX) && !defined(OS_LINUX)
 
 namespace quarantine {
 
@@ -19,4 +19,4 @@
 
 }  // namespace quarantine
 
-#endif  // !WIN && !MAC
+#endif  // !WIN && !MAC && !LINUX
diff -urN a/components/services/quarantine/test_support.cc b/components/services/quarantine/test_support.cc
--- a/components/services/quarantine/test_support.cc	2019-10-27 01:17:57.328896745 +0000
+++ b/components/services/quarantine/test_support.cc	2019-10-27 01:21:51.125965714 +0000
@@ -6,7 +6,7 @@
 
 #include "build/build_config.h"
 
-#if !defined(OS_WIN) && !defined(OS_MACOSX)
+#if !defined(OS_WIN) && !defined(OS_MACOSX) && !defined(OS_LINUX)
 
 namespace quarantine {
 
@@ -18,4 +18,4 @@
 
 }  // namespace quarantine
 
-#endif  // !WIN && !MAC
+#endif  // !WIN && !MAC && !LINUX
